### PR TITLE
Router metrics framework

### PR DESCRIPTION
**Pull request by pnarayanan**
Initial patch for the router metrics. This includes the framework, with
metrics common to all the operations. Additionally, a few metrics
specific to put and get operations are included. This will be followed
up with more metrics.

Primary reviewers: Ming.

Some of this overlaps with #326, and this PR is provided for reference for now. We are going to discuss offline and figure out how to resolve this, so please hold off on the review until then.

### CODE REVIEW

**Comment by pnarayanan** [comment]
We discussed and decided that we will have this PR go through first (after this gets reviewed and possibly removing some that is getting added as part of #326), and Ming will add more metrics to the non-blocking router in addition to #326.

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java [hunk 1]
```diff
@@ -148,6 +149,7 @@ void handleResponse(ResponseInfo responseInfo) {
     int correlationId = ((GetRequest) responseInfo.getRequest()).getCorrelationId();
     // Get the GetOperation that generated the request.
     GetRequestInfo getRequestInfo = correlationIdToGetRequestInfo.remove(correlationId);
+    routerMetrics.routerRequestLatencyMs.update(time.milliseconds() - getRequestInfo.startTimeMs);
```
Comment (xiahome) [inline ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java]: I am not very sure if `routerRequestLatencyMs` should be updated here. Are you measuring the time difference between start-to-process-response time and request-submitted time? It makes more sense to move this to the end of this method.

**Comment by xiahome** [comment]
I have gone through one pass, most look ok to me.

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java (hunk 1 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java]: This is the latency for a request. The start time is set at the time the request is handed over to the `NetworkClient`. The end time is set at the time the request is received from the `NetworkClient`. That way I think this is the right place.

**Comment by xiahome** [comment]
LGTM. This will be an initial PR, and from this point additions will be made to improve the router-metrics framework.

**Pull request merged**